pkg/library/gatewayapi/topology.go

```diff
@@ -0,0 +1,291 @@
+			// the parent gateway may not be in the available list of gateways
+			// or the gateway may not be valid
+			if gateway, ok := gatewaysIndex[parentKey]; ok {
+				edges = append(edges, edge{parent: gateway, child: route})
+			}
+		}
+	}
+
+	return edges
+}
+
+func buildGatewayDAGNodes(gateways []*gatewayapiv1.Gateway, policies []Policy) []gatewayDAGNode {
+	programmedGateways := utils.Filter(gateways, func(g *gatewayapiv1.Gateway) bool {
+		return meta.IsStatusConditionTrue(g.Status.Conditions, string(gatewayapiv1.GatewayConditionProgrammed))
+	})
+
+	return utils.Map(programmedGateways, func(g *gatewayapiv1.Gateway) gatewayDAGNode {
+		// Compute attached policies
+		attachedPolicies := utils.Filter(policies, func(p Policy) bool {
+			group := p.GetTargetRef().Group
+			kind := p.GetTargetRef().Kind
+			name := p.GetTargetRef().Name
+			namespace := ptr.Deref(p.GetTargetRef().Namespace, gatewayapiv1.Namespace(p.GetNamespace()))
+
+			return group == gatewayapiv1.GroupName &&
+				kind == "Gateway" &&
+				name == gatewayapiv1.ObjectName(g.Name) &&
+				namespace == gatewayapiv1.Namespace(g.Namespace)
+		})
+		return gatewayDAGNode{Gateway: g, attachedPolicies: attachedPolicies}
+	})
+}
+
+func buildHTTPRouteDAGNodes(routes []*gatewayapiv1.HTTPRoute, policies []Policy) []httpRouteDAGNode {
+	return utils.Map(routes, func(route *gatewayapiv1.HTTPRoute) httpRouteDAGNode {
+		// Compute attached policies
+		attachedPolicies := utils.Filter(policies, func(p Policy) bool {
+			group := p.GetTargetRef().Group
+			kind := p.GetTargetRef().Kind
+			name := p.GetTargetRef().Name
+			namespace := ptr.Deref(p.GetTargetRef().Namespace, gatewayapiv1.Namespace(p.GetNamespace()))
+
+			return group == gatewayapiv1.GroupName &&
+				kind == "HTTPRoute" &&
+				name == gatewayapiv1.ObjectName(route.Name) &&
+				namespace == gatewayapiv1.Namespace(route.Namespace)
+		})
+		return httpRouteDAGNode{HTTPRoute: route, attachedPolicies: attachedPolicies}
+	})
+}
+
+func (g *Topology) Gateways() []GatewayNode {
+	gatewayNodes := g.graph.GetNodes(typeField, gatewayLabel)
+
+	return utils.Map(gatewayNodes, func(n dag.Node) GatewayNode {
+		gNode, ok := n.(gatewayDAGNode)
+		if !ok { // should not happen
+			g.Logger.Error(
+				fmt.Errorf("node ID %s type %T", n.ID(), n),
+				"DAG gateway index returns nodes that are not gateways",
+			)
+			return GatewayNode{}
+		}
+
+		routeNodes := g.graph.Children(gNode.ID())
+		// convert to "RouteNode" from httpRouteDAGNode
+		routes := utils.Map(routeNodes, func(r dag.Node) RouteNode {
+			rDAGNode, ok := r.(httpRouteDAGNode)
+			if !ok { // should not happen
+				g.Logger.Error(
+					fmt.Errorf("node ID %s type %T", n.ID(), n),
+					"DAG index returns gateway children that are not routes",
+				)
+				return RouteNode{}
+			}
+			return RouteNode(rDAGNode)
+		})
+
+		return GatewayNode{
+			Gateway:          gNode.Gateway,
+			attachedPolicies: gNode.attachedPolicies,
+			routes:           routes,
```

Comment:
Routes of the gateways added lazily only when reading the gateways of a topology. Maybe add a comment?

Reply:
Sorry, I do not follow. When calling `Gateways()`, the returned object has all the routes for each gateway in the topology. It is not loaded lazily. There is no callback from the user perspective.